agent/lib/kontena/workers/fluentd_worker.rb

```diff
@@ -0,0 +1,60 @@
+require 'fluent-logger'
+
+module Kontena::Workers
+  class FluentdWorker
+    include Celluloid
+    include Celluloid::Notifications
+    include Kontena::Logging
+
+    attr_reader :fluentd, :node_name, :queue
+
+    ##
+    # @param [Queue] queue
+    # @param [Boolean] autostart
+    def initialize(autostart = true)
+      @fluentd = nil
+      @queue = []
+      @forwarding = false
+      info 'initialized'
+      subscribe('agent:node_info', :on_node_info)
+      subscribe('container:log', :on_log_event)
+    end
+
+    # @param [String] topic
+    # @param [Hash] info
+    def on_node_info(topic, info)
+      node_name = info['name']
+      driver = info.dig('grid', 'logs', 'driver')
+      if driver == 'fluentd'
+        fluentd_address = info.dig('grid', 'logs', 'opts', 'fluentd-address')
+        info "starting fluentd log streaming to #{fluentd_address}"
+        host, port = fluentd_address.split(':')
+        @fluentd = Fluent::Logger::FluentLogger
+          .new("#{node_name}.#{info.dig('grid' 'name')}",
+              :host => host,
+              :port => port || 24224,
+              :logger => Kontena::Logging.logger)
+        @forwarding = true
+      else
+        info "stopping fluentd log streaming"
+        @fluentd.close if @fluentd
+        @forwarding = false
+        @fluentd = nil
+      end
+    end
+
+    def on_log_event(topic, log)
+      # TODO Get more tags
+      # maybe stack.service.instance
+      #puts "forwarding: #{@forwarding}, fluentd: #{@fluentd}"
+      if @forwarding && @fluentd
+        tag = [log[:stack], log[:service], log[:instance]].join('.')
```

Comment:
No, the grid part is actually configured when instantiating the fluentd forwarder:
```
@fluentd = Fluent::Logger::FluentLogger
          .new("#{node_name}.#{info.dig('grid' 'name')}",
              :host => host,
              :port => port || 24224,
              :logger => Kontena::Logging.logger)
```